presto-testng-services/src/main/java/io/prestosql/testng/services/FlakyAnnotationVerifier.java

```diff
@@ -0,0 +1,124 @@
+import static io.prestosql.testng.services.Listeners.reportListenerFailure;
+import static java.util.stream.Collectors.joining;
+
+/**
+ * Detects test methods which are annotaded with @Flaky annotation but are
+ * missing explicit @Test annotation
+ */
+public class FlakyAnnotationVerifier
+        implements IClassListener
+{
+    @Override
+    public void onBeforeClass(ITestClass testClass)
+    {
+        try {
+            reportMethodsWithFlakyAndNoTestAnnotation(testClass);
+        }
+        catch (RuntimeException | Error e) {
+            reportListenerFailure(
+                    FlakyAnnotationVerifier.class,
+                    "Failed to process %s: \n%s",
+                    testClass,
+                    getStackTraceAsString(e));
+        }
+    }
+
+    private void reportMethodsWithFlakyAndNoTestAnnotation(ITestClass testClass)
+    {
+        Class<?> realClass = testClass.getRealClass();
+
+        if (realClass.getSuperclass() != null &&
+                "io.prestosql.tempto.internal.convention.ConventionBasedTestProxyGenerator$ConventionBasedTestProxy".equals(realClass.getSuperclass().getName())) {
+            // Ignore tempto generated convention tests.
+            return;
+        }
+
+        if (realClass.getName().startsWith("io.prestosql.testng.services.TestFlakyAnnotationVerifier")) {
+            // ignore test of FlakyAnnotationVerifier and internal classes
+            return;
+        }
+
+        List<Method> unannotatedTestMethods = findMethodsWithFlakyAndNoTestAnnotation(realClass);
+        if (!unannotatedTestMethods.isEmpty()) {
+            reportListenerFailure(
+                    FlakyAnnotationVerifier.class,
+                    "Test class %s has methods which are marked as @Flaky but are not explicitly annotated with @Test:%s",
+                    realClass.getName(),
+                    unannotatedTestMethods.stream()
+                            .map(Method::toString)
+                            .collect(joining("\n\t\t", "\n\t\t", "")));
+        }
+    }
+
+    @VisibleForTesting
+    static List<Method> findMethodsWithFlakyAndNoTestAnnotation(Class<?> realClass)
+    {
+        return Arrays.stream(realClass.getMethods())
+                .filter(method -> hasOrInheritsAnnotation(method, Flaky.class))
+                .filter(method -> !hasAnnotation(method, Test.class))
+                .filter(method -> hasOrInheritsAnnotation(method, Test.class))
+                .collect(toImmutableList());
+    }
+
+    @Override
+    public void onAfterClass(ITestClass testClass) {}
+
+    private static boolean hasAnnotation(Method method, Class<? extends Annotation> annotationClass)
+    {
+        return method.getAnnotation(annotationClass) != null;
+    }
+
+    private static boolean hasOrInheritsAnnotation(Method method, Class<? extends Annotation> annotationClass)
+    {
+        Optional<Method> currentMethod = Optional.of(method);
```

Comment:
Using `Optional` for loop control is IMO an overkill. Using plain old reference yields to a more readable code.